### DISCUSSION TITLE
The Status of the Directions Module

### DISCUSSION THREAD

**Discussion by amjaliks** [maplibre/maplibre-navigation-ios#87] [General]
Hello!

What is the status of the Directions module? Is there a plan to bring it under the MapLibre umbrella as well? Currently, MapLibre Navigation depends on a seemingly third-party managed [package/repository](https://github.com/flitsmeister/mapbox-directions-swift). However, this is a crucial part of the overall experience with MapLibre Navigation.

I'm interested in adding support for Speed Limit annotations. This requires changes to both: the Directions and Navigation repositories.

**Comment by boldtrn** [comment]
In the last steering comitte meeting we talked about that and decided that for now it would be the best approach to update to the latest version from mapbox. I think @Patrick-Kladek was interested in this as well. If you would like to get involved, feel free to open a PR 😇. Speed limit would be a great addition and is in fact missing right now.

**Reply by amjaliks** [reply]
Thanks for the update!

I'm ready to help, especially with the speed limit annotations. Should the PR just update the dependencies to the original Mapbox Directions, or is there more to it?

Looking forward to your guidance!

**Reply by boldtrn** [reply]
Yes, I think this would be the first step. I assume there might have been some breaking changes (I am not sure though). So you might have to make adaptions here as well.

**Reply by Patrick-Kladek** [reply]
I actually started working on this yesterday by researching how to make this work. In our case we use Graphhopper for the directions. I think the best approach would be to mimic the graphhopper response to Mapbox's response, then we can simply use the new Mapbox Directions Framework - so basically the same thing as you.

@amjaliks did you start on this already? I think it would be a good idea to coordinate our work so we can finish this faster.

**Reply by amjaliks** [reply]
Not yet – I'm currently in the research phase and coordinating with my client on how to integrate speed limit information. Their API already returns routes in a Mapbox-compatible format.

**Reply by Patrick-Kladek** [reply]
Starting with MapboxDirections [v0.31](https://github.com/mapbox/mapbox-directions-swift/releases/tag/v0.31.0) objc support was removed. This could mean we loose objc for MapLibre-Navigation as well.

**Reply by boldtrn** [reply]
Thanks for the info @Patrick-Kladek. I am wondering if we still need ObjC support? Is this still expected by iOS libraries in 2024?

**Comment by michaelkirk** [comment]
I would be fine if we officially dropped support. I’m not convinced we actually support objc right now anyway (please correct me if you know of an objc reject using the lib). I know we’ve sprinkled around some @objc annotations and such, but the bridging header looks pretty implausible.On Jul 25, 2024, at 08:16, Robin ***@***.***> wrote:﻿
Thanks for the info @Patrick-Kladek. I am wondering if we still need ObjC support? Is this still expected by iOS libraries in 2024?

**Reply by michaelkirk** [reply]
I forgot we seem to have at least one active objc user — @nnhubbard might want to weigh in.

related: https://github.com/maplibre/maplibre-navigation-ios/issues/79

**Reply by nnhubbard** [reply]
I am only using my Objective-C App Delegate (too hard to migrate to Swift at the moment) to hook up the CarPlayManager and then am writing Swift for all of my other CarPlay code using my own Swift classes.

**Reply by nnhubbard** [reply]
That being said, I think if support was removed it might break my project. But I am not 100% sure.

**Comment by amjaliks** [comment]
So? What we have in the end? Should we move fast and break things? 🙂

**Reply by Patrick-Kladek** [reply]
Im currently in the process of preparing a PR that adds back the original (& newest version of)  Mapbox's MapboxDirections to MapLibre-Navigation. 

In a call with @boldtrn we've decided to remove the objc annotations. If a consumer still needs objc support he should use  MapLibre in swift and wrap the code to be objc compatible. Thats the same approach MapboxDirections has given.

**Reply by Patrick-Kladek** [reply]
But yeah basically move fast and break things. As this will be a breaking change we should therefore release it as a major version.

**Reply by amjaliks** [reply]
Thanks for the update! I'll wait until these changes are merged, and then we can decide how to proceed with the Speed Limits.

**Comment by Patrick-Kladek** [comment]
After 2 days of trying to get Maplibre-Navigation to compile with MapboxDirections, I'm stopping now. Initially, we planned to update MapboxDirections so we get speed limit information parsing for free, and then we would need to add our UI.

Sadly this turns out to be way more complicated and time-consuming than we initially thought, so we are now changing course and potentially adding speed limit parsing to the currently used, filtsmeister's fork https://github.com/flitsmeister/mapbox-directions-swift/

I still think it's a good idea to use Mapbox's version of MapboxDirections as the new Swift Interface is nicer and we get a regularly maintained package but that doesn't fit in our timeline.

If anyone wants to give it a shot feel free to continue from my PR here: https://github.com/maplibre/maplibre-navigation-ios/pull/89

**Reply by boldtrn** [reply]
I totally get that, I just like to add, the old version is IMHO also messy to work with :) - not sure if @amjaliks would like to give this another try?

**Reply by amjaliks** [reply]
@Patrick-Kladek, what is the current status of Speed Limit support? Have you started on it? Maybe already finished?

@boldtrn, I could give it another try if that's still the preferred approach.

**Reply by Patrick-Kladek** [reply]
I gave up on this approach as basically all of MapLibre would need code changes and I would probably break something in the process. 

Instead I'll try to add speed limit to the old version of MapboxNavigation (if I remember correctly is the filtsmeister fork). 

As I'm currently on holiday I won't be working on this in the next weeks. Maybe you want to give the first/initial idea another try

**Reply by michaelkirk** [reply]
In case  you missed it, I opened #91 as a first step towards updating MapboxDirections.
